tools/clean.sh

```diff
@@ -0,0 +1,109 @@
+#!/bin/sh
+# Shell script equivalent of Metamath CLEAN tool
+# Mario Carneiro 5-Jan-2022
+
+usage() {
+  cat << HELP
+CLEAN - Trim spaces and tabs on each line in a file; convert characters
+
+This command processes spaces and tabs in each line of a file
+according to the following subcommands:
+  D - Delete all spaces and tabs
+  B - Delete spaces and tabs at the beginning of each line
+  E - Delete spaces and tabs at the end of each line
+  R - Reduce multiple spaces and tabs to one space
+  Q - Do not alter characters in quotes (ignored by T and U)
+  T - (Tab) Convert spaces to equivalent tabs (unsupported)
+  U - (Untab) Convert tabs to equivalent spaces (unsupported)
+Some other subcommands are also available:
+  P - Trim parity (8th) bit from each character
+  G - Discard garbage characters CR,FF,ESC,BS
+  C - Convert to upper case
+  L - Convert to lower case
+  V - Convert VT220 screen print frame graphics to -,|,+ characters
+Subcommands are separated by spaces, e.g., B E R Q
+Syntax:  clean.sh subcmd subcmd ... < INFILE > OUTFILE
+HELP
+}
+
+if [ $# -eq 1 ] && [ "$1" = "-h" ]; then usage; exit; fi
+if [ $# -eq 0 ]; then usage; exit 1; fi
+
+for var in "$@"; do
+  case $var in
+    D | d) allDiscard=1;;
+    B | b) leadDiscard=1;;
+    E | e) trailDiscard=1;;
+    R | r) reduce=1;;
+    Q | q) quote=1;;
+    T | t) echo "T (tab) flag unsupported"; exit 2;;
+    U | u) echo "U (untab) flag unsupported"; exit 2;;
+    P | p) trim=1;;
+    G | g) discardCtrl=1;;
+    C | c) caseFlag=1;;
+    L | l) lowercase=1;;
+    V | v) screen=1;;
+  esac
+done
+
+awk \
+  -v trim=$trim \
+  -v allDiscard=$allDiscard \
+  -v discardCtrl=$discardCtrl \
+  -v leadDiscard=$leadDiscard \
+  -v reduce=$reduce \
+  -v caseFlag=$caseFlag \
+  -v trailDiscard=$trailDiscard \
+  -v quote=$quote \
+  -v lowercase=$lowercase \
+  -v tab=$tab \
+  -v untab=$untab \
+  -v screen=$screen \
+'BEGIN {
+  FS="";
+  if (trim || caseFlag || lowercase) {
+    for (n=0; n<256; n++) {
+      ch = sprintf("%c", n);
+      chr[n] = ch;
+      ord[ch] = n;
+    }
+  }
+  if (discardCtrl) {
+    discardArr["\r"] = 1;
+    discardArr["\n"] = 1;
+    discardArr["\f"] = 1;
+    discardArr["\x27"] = 1;
+    discardArr["\x08"] = 1;
+  }
+  if (screen) {
+    gfxArr["\xEA"] = gfxArr["\xEB"] = gfxArr["\xEC"] = gfxArr["\xED"] = "+";
+    gfxArr["\xDA"] = gfxArr["\xD9"] = gfxArr["\xBF"] = gfxArr["\xC0"] = "+";
+    gfxArr["\xF1"] = gfxArr["\xC4"] = "-";
+    gfxArr["\xF8"] = gfxArr["\xA6"] = gfxArr["\xB3"] = "|";
+  }
+} {
+  if (leadDiscard)
+    gsub(/^[ \t]+/, "");
+  inQuote = 0;
+  out = "";
+  for (i=1;i<=NF;i++) {
+    if (quote && ($i == "\"" || $i == "\x39"))
+      inQuote = !inQuote;
+    if (inQuote) {
+      out=out $i;
+      continue;
+    }
+    if (allDiscard && ($i == " " || $i == "\t")) continue;
+    if (trim) $i = chr[and(ord[$i], 0x7f)];
+    if (discardCtrl && discardArr[$i]) continue;
```

Comment:
optimization: move this one line up

Reply:
That does not preserve behavior: if `trim` and `discardCtrl` are both enabled, then `0x88` would be discarded with the current order but not if the discard precedes the trim. This is a direct transcription from the `edit()` function in metamath so I would want to change it there too if so.

Reply:
There is a deeper problem here: If the parity bit is used to ensure proper transmission from an unreliable source, this parity bit clearance has to be executed first.  The space character SP 0x20, for example, is transmitted as 0xA0 if parity must be even.  This immediately busts the checks in line 86 and 96.  Bad.  So the parity bit has highest priority and line 97 must be moved upwards to line 90.  And line 86 is in a wrong position, too.  Why should leading spaces be handled differently from trailing ones?  If any of the transformations in the loop is capable of producing a space,  then the first character may be converted into a leading space,  which might be prohibited, and the algorithm fails again.

To get back to your remark: option P and V are mutually exclusive.  And who on earth cares about teletyping and non-graphical monitors any more?  So lets leave this issue at rest.  These options are eligible to removal.

Keeping the code for leading and trailing spaces in different locations still looks awkward.  But spaces or tabs are not produced by any transformation, so the implementation is somewhat risky, but not exactly failing.

That said I leave it up to you whether this is corrected.

Reply:
Like you say, most of the features here are significantly out of date and useless. My main interest is in preserving the behavior of the `edit()` function as written, at least for those options that can be simply implemented. I would very much like to remove the function entirely (from metamath) at some point, but that relies on having an appropriate fallback for applications that are still using it for some reason. So even though the behavior of this function is totally nonsensical when it comes to parity bit mangling (when was the last time the 8th bit was parity anyway?), I'm inclined to keep it as is for compatibility's sake.

Reply:
The parity flag is still present on super-modern Intel CPUs.  Its an heritage of the venerable 8086.  This roughly gives an idea when parity was used in transmission protocols